runtime/Tracer.cpp

```diff
@@ -0,0 +1,113 @@
+
+#include "Tracer.h"
+#include "Shadow.h"
+#include "llvm/Support/raw_ostream.h"
+
+nlohmann::json Tracer::currentTrace;
+nlohmann::json Tracer::expressions;
+nlohmann::json Tracer::pathConstraints;
+const std::string Tracer::BACKEND_TRACE_FILE = "/tmp/backend_trace.json";
+
+void Tracer::traceStep(uintptr_t pc) {
+
+  nlohmann::json newEntry;
+  newEntry["pc"] = pc;
+  newEntry["memory_to_expression_mapping"] = nlohmann::json::object();
+
+  /* dump shadow pages */
+  for (auto const &[pageAddress, _] : g_shadow_pages) {
+    for (auto byteAddress = pageAddress; byteAddress < pageAddress + kPageSize;
+         byteAddress++) {
+      auto byteExpr = _sym_read_memory((u_int8_t *)byteAddress, 1, true);
+      if (byteExpr != nullptr && !byteExpr->isConcrete()) {
```

Comment:
Maybe we can compare the current memory state with previous one to don't duplicate information in each step and only print diffs? This will reduce number of IO operations and needed memory.

Reply:
This would certainly be a useful improvement but unfortunately I will not have time to work on this.